portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -1,143 +1,155 @@
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Query query = new Query(QUOTE).addSort(TIMESTAMP, SortDirection.DESCENDING);
-    PreparedQuery quotes = datastore.prepare(query);
+    response.setContentType("application/json");
 
-    // Get the parameter for number to display, with 5 as the default value.
-    int numQuoteToDisplay = getNumQuoteDisplayed(request, DEFAULT_NUM_QUOTES);
-   
-    // Add each quote to the quote list whose content will be written as 
-    // the response from the servlet. 
-    List<Quote> quoteList = new ArrayList<>();
-    quotes.asList(FetchOptions.Builder.withLimit(numQuoteToDisplay))
-        .forEach((quoteEntity) -> {
-          quoteList.add(extractQuoteFromEntity(quoteEntity));
-        });
-    
-    // Put the thread to sleep for a short period of time.
-    // This adds buffer time for querying Datastore. 
-    try {
-      Thread.sleep(200);
-    } catch (Exception e) {
-      System.out.println(e);
+    UserService userService = UserServiceFactory.getUserService();
+    HashMap<String, String> responseMap = new HashMap<String, String>();
+
+    if (userService.isUserLoggedIn()) {
+      String userEmail = userService.getCurrentUser().getEmail();
+      String urlToRedirectToForLogOut = "/";
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToForLogOut);
+      String nickname = getUserNickname(userService.getCurrentUser().getUserId());  
+
+      responseMap.put(LOGGEDIN, "true");
+      responseMap.put(USEREMAIL, userEmail);
+      responseMap.put(REDIRECTURL, logoutUrl); 
+      responseMap.put(NICKNAME, nickname); 
+
+      // Only display quotes if the user is logged in
+      String quotesJson = getQuoteListJson(request);
+      responseMap.put(QUOTE, quotesJson);
+    } else {
+      String urlToRedirectToForLogIn = "/";
+      String loginUrl = userService.createLoginURL(urlToRedirectToForLogIn);
+      responseMap.put(LOGGEDIN, "false");
+      responseMap.put(USEREMAIL, null);
+      responseMap.put(REDIRECTURL, loginUrl); 
+      responseMap.put(NICKNAME, null); 
+      responseMap.put(QUOTE, null);
     }
 
-    String quotesJson = convertToJson(quoteList);
+    Gson gson = new Gson();
+    response.getWriter().println(gson.toJson(responseMap)); 
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json");
-    response.getWriter().println(quotesJson); 
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      String userEmail = userService.getCurrentUser().getEmail();
+      String nickname = getUserNickname(userService.getCurrentUser().getUserId()); 
+
+      // HashMap that contains user information: email address, nickname.
+      HashMap<String, String> userInfo = new HashMap<String, String>();
```

Comment:
Similarly, we can use the interface as the type.

Map<...>

Reply:
Done.